lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,168 @@
+    >;
+    verifyDevice: (input: {
+        userId: string;
+        deviceName: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; wasAlreadyVerified: boolean }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+        | { status: "INVALID_TOTP_ERROR" | "UNKNOWN_DEVICE_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    updateDevice: (input: {
+        userId: string;
+        existingDeviceName: string;
+        newDeviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "TOTP_NOT_ENABLED_ERROR" | "DEVICE_ALREADY_EXISTS_ERROR" | "UNKNOWN_DEVICE_ERROR" }>;
+    removeDevice: (input: {
+        userId: string;
+        deviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK"; didDeviceExist: boolean } | { status: "TOTP_NOT_ENABLED_ERROR" }>;
+    listDevices: (input: {
+        userId: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; devices: { name: string; period: number; skew: number; verified: boolean }[] }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+    >;
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    appInfo: NormalisedAppinfo;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    createDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
+        period?: number;
+        skew?: number;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; issuerName: string; secret: string; userIdentifier?: string; qrCode: string }
+        | { status: "DEVICE_ALREADY_EXISTS_ERROR" }
+    >;
+    verifyCodePOST?: (input: {
+        session: SessionContainer;
+        totp: string;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "INVALID_TOTP_ERROR" } | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }>;
+    verifyDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
+        totp: string;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; wasAlreadyVerified: boolean }
+        | { status: "INVALID_TOTP_ERROR" | "UNKNOWN_DEVICE_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    removeDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<{ status: "OK"; didDeviceExist: boolean }>;
+    listDevicesGET?: (input: {
+        session: SessionContainer;
+        options: APIOptions;
+        userContext: any;
+    }) => Promise<{ status: "OK"; devices: { name: string; period: number; skew: number; verified: boolean }[] }>;
```

Comment:
this API should return TOTP_NOT_ENABLED_ERROR as well.

Reply:
@KShivendu missed this?

Reply:
https://github.com/supertokens/supertokens-node/pull/576/files#r1223923156

Reply:
403 is a very specific error - reserved for claim validation failures. So we can't reuse that here. Instead, we should return `TOTP_NOT_ENABLED_ERROR`

Reply:
We decided to remove `TOTP_NOT_ENABLED_ERROR` from the core response. So this doesn't apply anymore.